### PR TITLE
improves numerical stability (see #376)

### PR DESCRIPTION
**Pull request by inkrement**
This request contains the following updates:

- solves numerical instabilities of adagrad implementation
- refactors sigmoid into own function & use a more stable implementation
- provide an updated version of the cython outputs (for people which directly call setup.py install)

### CODE REVIEW

File: implicit/lmf.pyx [hunk 1]
```diff
@@ -239,26 +246,26 @@ def lmf_update(RNGVector rng, floating[:, :] deriv_sum_sq,
...
 
+                    for _ in range(n_factors):
+                        deriv[_] = deriv[_] - z * item_vectors[i, _]
                 for _ in range(n_factors):
                     deriv[_] -= reg * user_vectors[u, _]
                     deriv_sum_sq[u, _] += deriv[_] * deriv[_]
-                    user_vectors[u, _] += (lr / sqrt(deriv_sum_sq[u, _])) * deriv[_]
+
+                    # a small constant is added for numerical stability
+                    user_vectors[u, _] += (lr / (sqrt(1e-6 + deriv_sum_sq[u, _]))) * deriv[_]
```
Comment (ita9naiwa) [inline implicit/lmf.pyx:269]: I think this will be better to change at line 171-172
```python
        user_vec_deriv_sum = np.zeros((users, self.factors + 2)).astype(np.float32)
        item_vec_deriv_sum = np.zeros((items, self.factors + 2)).astype(np.float32)
```

to
```python
        user_vec_deriv_sum = np.ones((users, self.factors + 2)).astype(np.float32) * EPS
        item_vec_deriv_sum = np.ones((items, self.factors + 2)).astype(np.float32) * EPS
```
where EPS is some value like 1e-3, 1, 1e-6

Comment (ita9naiwa) [inline implicit/lmf.pyx:269]: this line would be better

```python
    user_vectors[u, _] += (lr / (1 + sqrt(1e-6+ deriv_sum_sq[u, _]))) * deriv[_]
```

Comment (ita9naiwa) [reply implicit/lmf.pyx:269]: not 1, but 1.0

**Comment by ita9naiwa** [comment]
It seems Travis is not happy with this..

https://travis-ci.org/github/benfred/implicit/builds/714259293

File: implicit/lmf.pyx (hunk 1 above)
Comment (inkrement) [reply implicit/lmf.pyx:269]: Why +1 ? I just added a small number (an "epsilon"), adding a higher constant such as 1 would bias it towards smaller learning rates or not?

**Comment by inkrement** [comment]
The linter warnings are fixed now.

File: implicit/lmf.pyx (hunk 1 above)
Comment (ita9naiwa) [reply implicit/lmf.pyx:270]: yes but original paper suggests tha(cant find original paper link) but. https://www.slideshare.net/mobile/ssuser77b8c6/an-overview-of-gradient-descent-optimization-algorithms


I think it makes sense because if we dont add 1, then learning rate at first few steps becomes extremely large because it divides sqrt of sum of square of grads(nearly 0 mostly)

Comment (ita9naiwa) [reply implicit/lmf.pyx:270]: making learning rate larger than what we initialized

**Comment by inkrement** [comment]
Linting warnings are fixed now.

File: implicit/lmf.pyx (hunk 1 above)
Comment (inkrement) [reply implicit/lmf.pyx:270]: I still can't follow you: IMHO 1 is *not* a small constant (especially, for tuning a learning rate). Even your link suggests 1e-8 (=0.00000001) as constant and does not mention the addition of 1 at all. I also looked at other sources (e.g., https://d2l.ai/chapter_optimization/adagrad.html) and there they also use a constant of 1e-6. I would understand it to use a smaller one than 1e-6, but 1.0 is one million times higher than other sources suggest.

Comment (ita9naiwa) [reply implicit/lmf.pyx:270]: You are right, 1 is super larger number...

https://arxiv.org/pdf/1609.04747 suggests 1e-8

the default implementation in PyTorch suggests 1-e10
https://pytorch.org/docs/stable/optim.html?highlight=adagrad#torch.optim.Adagrad

btw, those implementations place epsilon out of square root and I'm not sure putting epsilon inside of sqrt is good or not(anyway would be much better before)

**Comment by inkrement** [comment]
I just saw, that you started (and closed?) another discussion about the epsilon. IMHO, putting the epsilon in the beginning of the function is a bad idea because of 1) then its is only applied once, but the problem may occur also in later steps and 2) people used to adagrad implementations may overlook it. Especially, the second point should be considered carefully. 

The current status quo works well for my use case and is inline with other implementations. This is clearly a step forward in the right direction and I kindly ask you to provide these fixes asap. We should not forget, that the current version is highly instable and does not work properly in many situations. Given this context, discussions about the size and placement of an epsilon should be postponed to future enhancements.

**Comment by ita9naiwa** [comment]
I agree with your opinion totally.

but travis api has a problem so i cannot merge this now.

**Comment by inkrement** [comment]
Strange.. still seems to hang somewhat. I may add some minor changes to force another run. Is it ok, if I update the version to 0.4.3? It's anyhow strange that it is still 0.4.0, while there is a 0.4.2 on pypi or is there a reason for this?

**Comment by ita9naiwa** [comment]
This error is not related your work... seems becausebtravis api has changed and this repo didnt updated.

**Comment by ita9naiwa** [comment]
i sent email to owner(@benfred) of thisbrepo to check CI status. will be handled soon.

**Comment by BartyzalRadek** [comment]
Hi, sorry to hijack this thread a little bit, but are you planning to release these changes on PyPi as well? 
This library is awesome and we would like to use some of the new commits in master that are not on PyPi.
Is there some reason why is not the version updated there?

**Comment by Declow** [comment]
@BartyzalRadek
You can install the master branch to your machine using pip
pip install -e git+https://github.com/benfred/implicit.git@master#egg=implicit
